src/replication.c

```diff
@@ -775,31 +1048,101 @@ get_non_quorum_replica_count(spec_t *spec)
 static bool
 can_replica_connect(spec_t *spec, replica_t *replica)
 {
-	int replica_count, non_quorum_count;
+	int replica_count, non_quorum_count, trusty_replica_count;
 	bool ret = true;
 
 	ASSERT(MTX_LOCKED(&spec->rq_mtx));
-
+	non_quorum_count = get_non_quorum_replica_count(spec);
+
+	//TODO: Change non_quorum variables to unknown_quorum
+	//TODO: If DRF unknown replicas are already connected to the target.
+	//      If known replica came to connect we need to remove connected
+	//      replicas(in the order non-quorum, not-in-the-known-list)
+
+	/* We need to allow DRF replicas to connect to target there might
+	 * be case where rebuilding is done at replica and marked quorum=on,
+	 * immediately after that target or replica is restarted,
+	 * in this scenario we need to allow replica to connect and update
+	 * in non-quorum list and another use case in replica scale up scenarios.
+	 */
 	replica_count = spec->healthy_rcount + spec->degraded_rcount;
-	if (replica_count >= spec->replication_factor) {
+	if ((replica_count >= spec->desired_replication_factor) ||
+	    ((replica_count + non_quorum_count) >= spec->desired_replication_factor)) {
 		REPLICA_ERRLOG("removing replica(%s:%d) since healthy: %d"
-		    " degraded: %d connected >= (%d)", replica->ip,
-		    replica->port, spec->healthy_rcount,
-		    spec->degraded_rcount, spec->replication_factor);
-		ret = false;
-	} else {
-		non_quorum_count = get_non_quorum_replica_count(spec);
-		if ((non_quorum_count + replica_count) >= MAXREPLICA) {
-			REPLICA_ERRLOG("erroring replica(%s:%d) since "
-			    "non_quorum: %d healthy: %d degraded: %d are"
-			    " connected >= %d", replica->ip, replica->port,
-			    non_quorum_count, spec->healthy_rcount,
-			    spec->degraded_rcount, MAXREPLICA);
-			ret = false;
-		}
+		    " degraded: %d non_quorum: %d connected >= (%d)\n",
+		    replica->ip, replica->port, spec->healthy_rcount,
+		    spec->degraded_rcount, non_quorum_count,
+		    spec->desired_replication_factor);
+		return false;
 	}
 
-	return ret;
+	/* known and trusty replica */
+ 	ret = check_is_it_trusty_replica(spec, replica);
+	if (ret == true)
+		return true;
+
+	/* replica replacement case */
+	ret = is_trusted_replicas_contain_replicaid(spec, replica->replica_id);
+	if (ret == true)
+		return true;
+
+	/*
+	 * Unknown replica and its ID (replica GUID may be known - but doesn't matter)
+	 * i.e., new volume case, or, upgrade from 1.2 to 1.3
+	 * or scaleup if RF < DRF
+	 */
+	/* Allow if trusty_count < RF */
+	trusty_replica_count = get_trusty_replica_count(spec);
+	if (trusty_replica_count < spec->replication_factor)
+		return true;
+
+	/*
+	 * Unknown replica and its ID
+	 * trusty count == RF
+	 * this can be scale up case if RF < DRF
+	 */
+	if (spec->replication_factor < spec->desired_replication_factor)
```

Comment:
Can we a single check `trusty_replica_count < spec->desired_replication_factor`?

Reply:
I think it will not work in below scenarios
1. Let us assume that all replicas are connected and istgt conf file is updated with trusty replica list (i.e trusty_replica_count == drf).
2. Now target container is restarted and it will read trusty replicas from conf file and this check will fail and it will not allow any replica to connect to the target.